### PR TITLE
feat: added speculative sampling

### LINKED ISSUE #199: Speculative Execution
Like andrej karpathy said: https://twitter.com/karpathy/status/1697318534555336961

### PR DESCRIPTION
**Pull request by us**
close #199

Do you have any idea why speculative sampling result is different? It gives completely random results in every run!
 I thought speculative sampling was an alternative to Optimizer, am I wrong? any suggestion, idea? Please let me know!

Normal Result:
```
'NoneType' object has no attribute 'cadam32bit_grad_fp32'
Project Root: /Users/syny/kafamagore/LLM-VM/src/llm_vm
Using model: opt
Running with an empty context
{'completion': '\n\nThe capital of the USA is Washington, DC.\n\nThe capital of the USA is Washington, DC.\n\nThe capital of the USA is Washington, DC.\n\nThe capital of the USA is Washington, DC.\n\nThe capital of the USA is Washington, DC.\n\nThe capital of the USA is Washington, DC.\n\nThe capital of the USA is Washington, DC.\n\nThe capital of the USA is Washington, DC.\n\nThe capital of the USA is Washington, DC.\n\nThe capital of the USA is Washington, DC.\n\nThe capital of the USA is Washington, DC.\n\nThe capital of the USA is Washington, DC.\n\nThe capital of the USA is Washington, DC.\n\nThe capital of the USA is Washington, DC.\n\nThe capital of the USA is Washington, DC.\n\nThe capital of the USA is Washington, DC.\n\nThe capital of the USA is Washington, DC.\n\nThe capital of the USA is Washington, DC.\n\nThe capital of the USA is Washington, DC.\n\nThe capital of the USA is Washington, DC.\n\nThe capital of the USA', 'status': 200}
```

Speculative Sampling result:
```
'NoneType' object has no attribute 'cadam32bit_grad_fp32'
/Users/syny/kafamagore/LLM-VM/.venv/lib/python3.10/site-packages/bitsandbytes/cextension.py:34: UserWarning: The installed version of bitsandbytes was compiled without GPU support. 8-bit optimizers, 8-bit multiplication, and GPU quantization are unavailable.
  warn("The installed version of bitsandbytes was compiled without GPU support. "
Project Root: /Users/syny/kafamagore/LLM-VM/src/llm_vm
Using model: opt
huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...
To disable this warning, you can either:
	- Avoid using `tokenizers` before the fork if possible
	- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)
speculative sampling: 100%|█████████▉| 264/265 [02:07<00:00,  2.07it/s]
{'completion': "What is the capital of the USA? (Gregsergio, first person to say that.)  Edit: The wikipedia says it's New York City, so actually Washington D.C.  Edit 2: Did you check the wiki? They have an indication as to how it formed.\nNew York City is the capital of the USA; parts of what you say are in New York, particularly the boroughs. Washington D. C., narrowly speaking isn't a part of Washington, DC; the District is either D.C. or New York.  After the last I checked, you are correct, that I didn't check the wiki, so thanks!\nSorry, based on the wikipedia form, I assumed that the crown prince of Washington would be the queen, since she was holding it together during the Dark Ages. But if you're talking about Danzig, it's not a part of D.C., which wasn't anything to do with the time period, but different time period. Is Jerusalem the capital of Israel? A friend in Israel is Jewish, and he said Jerusalem, not Jerusalem.\nOh God, now I get it. I start reading in HS again and I assumed in high school i was 7 at the time. Thanks a lot!Not a case of 'they", 'status': 200}
```

### CODE REVIEW

File: src/llm_vm/utils/speculative_sampling.py [hunk 1]
```diff
@@ -0,0 +1,290 @@
+"""
+Speculative Sampling
+from https://raw.githubusercontent.com/feifeibear/LLMSpeculativeSampling/main/sample.py
+"""
+from tqdm import tqdm
```
Comment (cartazio) [inline src/llm_vm/utils/speculative_sampling.py:5]: what package is this? do we already depend on it?

File: src/llm_vm/utils/speculative_sampling.py [hunk 2]
```diff
@@ -0,0 +1,290 @@
+"""
+Speculative Sampling
+from https://raw.githubusercontent.com/feifeibear/LLMSpeculativeSampling/main/sample.py
+"""
```
Comment (mmirman) [inline src/llm_vm/utils/speculative_sampling.py:4]: investigating why this doesn't work seems like an important task

File: src/llm_vm/utils/speculative_sampling.py (hunk 1 above)
Comment (us) [reply src/llm_vm/utils/speculative_sampling.py:5]: it's a progress bar package, we already have! I can replace it if it needed!

and fun fact about the name 😊 [tqdm derives from the Arabic word taqaddum (تقدّم) which can mean "progress," and is an abbreviation for "I love you so much" in Spanish (te quiero demasiado).](https://github.com/tqdm/tqdm#:~:text=tqdm%20derives%20from%20the%20Arabic%20word%20taqaddum%20(%D8%AA%D9%82%D8%AF%D9%91%D9%85)%20which%20can%20mean%20%22progress%2C%22%20and%20is%20an%20abbreviation%20for%20%22I%20love%20you%20so%20much%22%20in%20Spanish%20(te%20quiero%20demasiado).)

**Comment by us** [comment]
- Updated `speculative_sampling.py` from main code.
	- 2023.09.19 Add KVCache Optimization to the Google's version.
- the issue of differences was `temperature` parameter, `0` gives this error, `status': 0, 'resp': 'probability tensor contains either `inf`, `nan` or element < 0` so I was used 1 and it creates really different results, now I tried with smaller temperature, first I tried with `0.1`, it's also gives same error after some steps, then I tried with `0.2` and works well. Here is the current output: 
	-`speculative sampling: 266it [01:24,  3.15it/s] {'completion': 'What is the capital of the USA?\n\nThe capital of the United States is Washington, DC.\n\nThe capital of the United States is Washington, DC.\n\nThe capital of the United States is Washington, DC.\n\nThe capital of the United States is Washington, DC.\n\nThe capital of the United States is Washington, DC.\n\nThe capital of the United States is Washington, DC.\n\nThe capital of the United States is Washington, DC.\n\nThe capital of the United States is Washington, DC.\n\nThe capital of the United States is Washington, DC.\n\nThe capital of the United States is Washington, DC.\n\nThe capital of the United States is Washington, DC.\n\nThe capital of the United States is Washington, DC.\n\nThe capital of the United States is Washington, DC.\n\nThe capital of the United States is Washington, DC.\n\nThe capital of the United States is Washington, DC.\n\nThe capital of the United States is Washington, DC.\n\nThe capital of the United States is Washington, DC.\n\nThe capital of the United States is Washington, DC.\n\nThe capital of the United States is Washington, DC.\n\nThe capital of the United States is Washington, DC', 'status': 200}`

File: src/llm_vm/client.py [hunk 3]
```diff
@@ -109,7 +127,12 @@ def complete(self, prompt,
         static_context = context
         dynamic_prompt = prompt
         use_rebel_agent = False
-
+        # TODO(us): speculative_sampling gives error when temp is 0,
+        # self.optimizer is None means `speculative_sampling` is used
```
Comment (cartazio) [inline src/llm_vm/client.py:131]: Shouldn’t we add an explicit flag for speculation?

Comment (us) [reply src/llm_vm/client.py:131]: in the current code, as default mode, speculative sampling and as optional normal optimizer can be used, as default should I keep the current optimizer?

**Comment by VictorOdede** [comment]
this can now be done using assisted generation on HF models

**Pull request closed without merging**